src/nft_contract/src/main.nr

```diff
@@ -50,118 +47,120 @@ pub contract NFT {
         minter: PublicImmutable<AztecAddress, Context>,
     }
 
+    /// @dev Initializes the token with a minter
+    /// @param name The name of the NFT
+    /// @param symbol The symbol of the NFT
+    /// @param minter The address of the minter
     #[public]
     #[initializer]
     fn constructor_with_minter(name: str<31>, symbol: str<31>, minter: AztecAddress) {
         storage.name.initialize(FieldCompressedString::from_string(name));
         storage.symbol.initialize(FieldCompressedString::from_string(symbol));
-
         storage.minter.initialize(minter);
     }
 
     /** ==========================================================
      * ========================= PRIVATE =========================
      * ======================================================== */
 
-    // Transfers token with `token_id` from public balance of message sender to a private balance of `to`.
     #[private]
-    fn transfer_to_private(to: AztecAddress, token_id: Field) {
-        let from = context.msg_sender();
-
-        let nft = NFT::at(context.this_address());
-
-        // We prepare the private balance increase.
-        let commitment = _prepare_private_balance_increase(to, &mut context, storage);
-
-        // At last we finalize the transfer. Usage of the `unsafe` method here is safe because we set the `from`
-        // function argument to a message sender, guaranteeing that he can transfer only his own NFTs.
-        nft._finalize_transfer_to_private_unsafe(from, token_id, commitment).enqueue(&mut context);
+    fn transfer_private_to_public(from: AztecAddress, to: AztecAddress, token_id: Field, nonce: Field) {
+        _validate_from_private(&mut context, from, nonce);
+        _remove_private_token_owner(storage.private_nfts.at(from), token_id);
+        NFT::at(context.this_address()).update_public_token_owner_internal(from, to, token_id, true).enqueue(
+            &mut context,
+        );
     }
 
-    /// Prepares an increase of private balance of `to` (commitment). The increase needs to be finalized by calling
-    /// `finalize_transfer_to_private` with the returned commitment.
     #[private]
-    fn prepare_private_balance_increase(to: AztecAddress) -> PartialNFTNote {
-        _prepare_private_balance_increase(to, &mut context, storage)
+    fn transfer_private_to_public_with_commitment(from: AztecAddress, to: AztecAddress, token_id: Field, commitment: PartialNFTNote, nonce: Field) -> PartialNFTNote {
+        _validate_from_private(&mut context, from, nonce);
+        _remove_private_token_owner(storage.private_nfts.at(from), token_id);
+        NFT::at(context.this_address()).update_public_token_owner_internal(from, to, token_id, true).enqueue(
+            &mut context,
+        );
+        let commitment = _initialize_transfer_commitment(&mut context, storage.private_nfts, from, to);
+        commitment
```

Comment:
i think we should deprecate this one as there's no visible usecase, but let's address that separatedly

Reply:
my bad, there might be one 🙃